### PR TITLE
Working on a S3 backend for map storage

### PR DESCRIPTION
**Pull request by moufmouf**
This is 100% WIP. Do not merge!

- [x] S3 Storage
- [x] S3 upload
- [x] Disk upload
- [x] Support for subdomains
- [x] Add regex to make files with name `xxx.[hash].[js|png|css|...]` unmutable in CacheControl / Configure CacheControl on DiskFileSystem
- [x] Support for directories in upload
- [x] Tests (Playwright API Tests?)
- [x] Documentation 
- [x] Authentication
- [x] Lock the upload route for a given path until previous request is done (can be done with a p-limit of 1)
- [x] Download endpoint
- [x] End-to-end test for download point (upload zip, download a subdir, reupload the subdir in another dir, access the file)
- [ ] Find a way to upload default files on startup in dev
- [x] An HTML page with a simple form to do an upload
- [x] Fix file reading / writing for *.json endpoint (need to take the domain name into account)
- [x] Look for a proper eslint addon for express (?)

### CODE REVIEW

**Comment by Waclaw-I** [comment]
Apart from CodeQL checks, LGTM! 👌

File: map-storage/src/Services/S3Client.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+export function getS3Client(): S3 {
+    if (s3) {
+        return s3;
+    }
+
+    const config: S3ClientConfig = {
+        // ID and Secret are detected by default by the lib
+        /*accessKeyId: AWS_ACCESS_KEY_ID,
+        secretAccessKey: AWS_SECRET_ACCESS_KEY,*/
```
Comment (gparant) [inline map-storage/src/Services/S3Client.ts:22]: Normal that this code is commented?

If it's useless, delete it

File: map-storage/src/Upload/S3FileSystem.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                }
+
+                // TODO assess this idea: REDIRECT TO THE BUCKET PUBLIC URL FOR MOST FILES EXCEPT .html files
+
+                if (!result.Body) {
+                    throw new Error("Missing body");
+                }
+                // Stream the item to the client
+                //(result.Body as ReadableStream).pipeTo(res);
+                //console.log('BAA', await result.Body.transformToByteArray());
```
Comment (gparant) [inline map-storage/src/Upload/S3FileSystem.ts:101]: delete comment useless

File: map-storage/src/Upload/S3FileSystem.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                if (result.ContentLength) {
+                    res.set("Content-Length", result.ContentLength.toString());
+                }
+
+                if (result.ETag) {
+                    res.set("ETag", result.ETag);
+                }
+
+                /*if (result.CacheControl) {
+                res.set('Cache-Control', result.CacheControl);
```
Comment (gparant) [inline map-storage/src/Upload/S3FileSystem.ts:87]: delete comment useless

File: map-storage/src/Upload/S3FileSystem.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                //(result.Body as ReadableStream).pipeTo(res);
+                //console.log('BAA', await result.Body.transformToByteArray());
+
+                // Typescript doc is wrong in AWS: see: https://github.com/aws/aws-sdk-js-v3/issues/1877#issuecomment-755387549
+                //eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                //@ts-ignore
+                (result.Body as IncomingMessage).pipe(res);
+                //console.log(result.Body.constructor.name);
+                //res.send(result.Body.transformToWebStream());
+                //res.end();
```
Comment (gparant) [inline map-storage/src/Upload/S3FileSystem.ts:109]: delete comment useless

File: map-storage/src/Upload/UploadController.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            });
+
+            // good practice to catch warnings (ie stat failures and other non-blocking errors)
+            archive.on("warning", function (err) {
+                if (err.code === "ENOENT") {
+                    // log warning
+                    console.warn("File not found: ", err);
+                } else {
+                    console.error("An warning occurred while Zipping file: ", err);
+                    //res.status(500).send('An error occurred');
```
Comment (gparant) [inline map-storage/src/Upload/UploadController.ts:174]: If there are warning, we never return status or body?

File: map-storage/src/index.ts [hunk 6]
```diff
@@ -22,15 +27,23 @@ server.bindAsync(`0.0.0.0:50053`, grpc.ServerCredentials.createInsecure(), (err,
 const app = express();
 app.use(cors());
 
-app.get("/", (req, res) => {
-    res.send("Hello World!");
-});
+passport.use(passportStrategy);
+app.use(passport.initialize());
+//app.use(passport.session());
```
Comment (gparant) [inline map-storage/src/index.ts:32]: If useless, delete it

File: map-storage/src/index.ts [hunk 7]
```diff
@@ -22,15 +27,23 @@ server.bindAsync(`0.0.0.0:50053`, grpc.ServerCredentials.createInsecure(), (err,
...
 
 app.get("*.json", async (req, res) => {
-    res.send(await mapsManager.getMap(req.url));
+    const path = req.url;
+    const domain = req.hostname;
+    if (path.includes("..") || domain.includes("..")) {
+        res.status(400).send("Invalid request");
+        return;
+    }
+    res.send(await mapsManager.getMap(path, domain));
```
Comment (gparant) [inline map-storage/src/index.ts:41]: If there are error in `mapsManager`, what's happening? No data returned?

Might we manage error et return status with body?

File: map-storage/src/Services/S3Client.ts (hunk 1 above)
Comment (moufmouf) [reply map-storage/src/Services/S3Client.ts:22]: Done

File: map-storage/src/Upload/UploadController.ts (hunk 5 above)
Comment (moufmouf) [reply map-storage/src/Upload/UploadController.ts:174]: I'm not sure we can do that. I'm "streaming" the ZIP file so I'm starting sending the answer BEFORE the full ZIP file is generated. We could have an issue in the middle of the ZIP. At this point, it is too late to change the HTTP status...

File: map-storage/src/index.ts (hunk 7 above)
Comment (moufmouf) [reply map-storage/src/index.ts:41]: Oooooh good point!
I put an exception just aboe to see how Express behaves when not catching an error.....

The result? NodeJS crashes :O

**Pull request merged**